File: KMACTF 2024/ZKP for QNR/python/prover.py
import random
from typing import List

class Prover:
    def __init__(self, x: int, y: int, n: int):
        self.x = x
        self.y = y
        self.num_rounds = n
        self.list_i = []

    def gen_list_i(self) -> List[int]:
        list_i = [random.randint(0, 1) for _ in range(self.num_rounds)]
        return list_i

File: KMACTF 2024/ZKP for QNR/src/prover.rs
use num_bigint::BigUint;
use rand::Rng;

pub struct Prover {
    pub x: BigUint,
    pub y: BigUint,
    pub num_rounds: usize,
    pub list_i: Vec<Vec<u32>>,
}

impl Prover {
    pub fn new(x: BigUint, y: BigUint, n: usize) -> Self {
        Prover {
            x,
            y,
            num_rounds: n,
            list_i: Vec::new(),
        }
    }

    pub fn gen_list_i(&self) -> Vec<u32> {
        let mut rng = rand::thread_rng();
        let list_i: Vec<u32> = (0..self.num_rounds).map(|_| rng.gen_range(0..2)).collect();
        list_i
    }

    // I forgot to implement verify function
    // pub fn verify() -> {
    //
    // }
}
